### PR TITLE
Feature: Case-insensitive "find files in repo"

### PR DESCRIPTION
**Pull request by HeySora**
This (short) PR builds upon #15028 and makes the file search case-insensitive.

Previously, having a file named `TestFile.cs` would not be shown if `test` was typed in the search box.  
This now changes the matching function to be case-insensitive (without affecting the UI).

The matching function, `strSubMatch`, is only used for this feature (it has been introduced by #15028), meaning that this PR does not affect the behaviour of any unrelated functionality of Gitea.

### CODE REVIEW

File: web_src/js/utils.js [hunk 1]
```diff
@@ -66,7 +66,7 @@ export function strSubMatch(full, sub) {
   let i = 0, j = 0;
   while (i < sub.length && j < full.length) {
     while (j < full.length) {
-      if (sub[i] === full[j]) {
+      if (typeof sub[i] !== 'undefined' && typeof full[j] !== 'undefined' && sub[i].toLowerCase() === full[j].toLowerCase()) {
```
Comment (KN4CK3R) [inline web_src/js/utils.js]: Why are the type checks necessary?

Comment (HeySora) [reply web_src/js/utils.js]: Because I noticed that `sub[i]` was undefined under some circumstances (and a call to `toLowerCase()` would thus error out.

I am guessing this could be due to the `i++` instruction on line 73, since nothing checks whether `i` is still within `sub`'s length or not.

I preferred staying on the safe side and do a type check on both variables instead of refactoring the loop and potentially making some edge case different.

Checking whether `i < sub.length` would effectively make the while loop on line 68 identical to the one on line 67.. which looks odd? Again, I preferred not refactoring the whole algorithm for this edge case, and instead type-checking this.

Would you like me to handle this differently?

Comment (silverwind) [reply web_src/js/utils.js]: Should just write `sub[i]?.toLowerCase() === full[j]?.toLowerCase()` if typecheck is really necessary, depends on how these arrays are generated. If they are guaranteed to be strings, the `?` is unnecessary.

Comment (HeySora) [reply web_src/js/utils.js]: `sub` and `full` are strings actually! the issue is that `i` can get too big, and in all honesty, I'm slightly confused to why a double-loop is necessary here at all.

`full` is for example "Folder/File.cs"  
`sub` is the query a user types into the "Find file" input field.

I'm OK doing the type-check as `?` instead. I did a Ctrl+F in the codebase to see how checks are done elsewhere, and went with `typeof` because that's what I found; i have no issues writing it as you suggested, if you're okay with it.

Comment (KN4CK3R) [reply web_src/js/utils.js]: > Checking whether i < sub.length would effectively make the while loop on line 68 identical to the one on line 67.. which looks odd? Again, I preferred not refactoring the whole algorithm for this edge case, and instead type-checking this.

Yes, the inner loop looks strange. If you remove it, one of the tests fails but maybe the test is wrong...

Comment (KN4CK3R) [reply web_src/js/utils.js]: works:
```
  while (i < sub.length && j < full.length) {
    while (j < full.length) {
```
does not work:
```
  while (i < sub.length && j < full.length) {
```
works and should not need type checks:
```
  while (i < sub.length && j < full.length) {
    while (i < sub.length && j < full.length) {
```
I will have a look with the debugger why the second one does not work.

Comment (HeySora) [reply web_src/js/utils.js]: The tests are correct! I think the current algorithm needs to loop over the full string multiple times (which happens when the inner loop is broken from line 78).

FWIW I just committed silverwind's suggestion (I completely forgot this was a thing in JS, my bad 😅)

Comment (KN4CK3R) [reply web_src/js/utils.js]: Yes, the break needs to get removed. Then there should be no typecheck needed.

Comment (KN4CK3R) [reply web_src/js/utils.js]:
```js
  while (i < sub.length && j < full.length) {
    if (sub[i].toLowerCase() === full[j].toLowerCase()) {
      if (res.length % 2 !== 0) res.push('');
        res[res.length - 1] += full[j];
        j++;
        i++;
      } else {
        if (res.length % 2 === 0) res.push('');
        res[res.length - 1] += full[j];
        j++;
      }
    }
  }
```

Comment (HeySora) [reply web_src/js/utils.js]: Done! I wonder why the code was like this in the first place...

**Review by KN4CK3R — APPROVED** [review]

**Comment by wxiaoguang** [comment]
Good catch!

ps: if we can have `subLower` and `fullLower` ahead, it would be faster.


the link [benchmark](https://perf.link/#eyJpZCI6Imh4eGFzeTAwYjFnIiwidGl0bGUiOiJ0ZXN0IHRvTG93ZXJDYXNlIiwiYmVmb3JlIjoiIiwidGVzdHMiOlt7Im5hbWUiOiJ3aG9sZSB0b0xvd2VyQ2FzZSIsImNvZGUiOiJsZXQgcmVzID0gMDtcbmxldCBzID0gJ2FiY2RBQkNEMTIzNGFiY2RBQkNEMTIzNGFiY2RBQkNEMTIzNCc7XG5jb25zdCB0ID0gcy50b0xvd2VyQ2FzZSgpO1xuZm9yIChsZXQgaSA9IDA7IGkgPCB0Lmxlbmd0aDsgaSsrKSB7XG4gIGlmICh0W2ldID09PSAnYScpIHJlcysrO1xufVxuY29uc29sZS5sb2cocmVzKTsiLCJydW5zIjpbMTI1MDAwLDg0MDAwLDcwMDAwLDEwMDAsMTE1MDAwLDE3MDAwLDQxMDAwLDc2MDAwLDcwMDAwLDY2MDAwLDEyNTAwMCwxODAwMCw4MDAwLDM1MDAwLDE1MDAwLDY1MDAwLDUxMDAwLDYxMDAwLDkxMDAwLDk0MDAwLDE4MDAwLDU1MDAwLDEyOTAwMCw0MDAwMCw5MDAwLDQ3MDAwLDEwMjAwMCw4NzAwMCwxMjYwMDAsMzAwMCwxMDAwMDAsNTIwMDAsNDAwMCwxMjkwMDAsNjAwMCwxOTAwMCw3NDAwMCw4OTAwMCwxMDAwMDAsOTgwMDAsMjQwMDAsNzkwMDAsMTM1MDAwLDU5MDAwLDExMzAwMCw3ODAwMCw5NzAwMCwxMDAwMCw2OTAwMCw0NjAwMCw3ODAwMCwxMDIwMDAsNzQwMDAsMjAwMDAsNzEwMDAsMTE3MDAwLDEwMDAsNjIwMDAsMTE0MDAwLDUzMDAwLDgyMDAwLDEzODAwMCwyMjAwMCwxMDAwLDExMzAwMCw0MTAwMCwxMDMwMDAsNTcwMDAsNzMwMDAsNjgwMDAsMzAwMDAsMzMwMDAsMTEzMDAwLDg1MDAwLDgwMDAwLDEyMDAwMCwxNDMwMDAsODYwMDAsOTEwMDAsOTQwMDAsMTc3MDAwLDQ4MDAwLDQ3MDAwLDEwMDAsMTEwMDAsODEwMDAsMzEwMDAsNTgwMDAsMTAxMDAwLDEwNzAwMCw1MzAwMCw1NDAwMCwzMDAwLDY3MDAwLDEzMDAwLDE0MDAwMCwxMDUwMDAsMzUwMDAsNzYwMDAsNDcwMDBdLCJvcHMiOjY3NDUwfSx7Im5hbWUiOiJlYWNoIHRvTG93ZXJDYXNlIiwiY29kZSI6ImxldCBzID0gJ2FiY2RBQkNEMTIzNGFiY2RBQkNEMTIzNGFiY2RBQkNEMTIzNCc7XG5sZXQgcmVzID0gMDtcbmZvciAobGV0IGkgPSAwOyBpIDwgcy5sZW5ndGg7IGkrKykge1xuICBpZiAoc1tpXS50b0xvd2VyQ2FzZSgpID09PSAnYScpIHJlcysrO1xufVxuY29uc29sZS5sb2cocmVzKTsiLCJydW5zIjpbMTcwMDAsODQwMDAsNDQwMDAsNjUwMDAsNzIwMDAsODAwMDAsMzYwMDAsNjAwMDAsMzIwMDAsMTAwMCw3MDAwMCw3MDAwMCw2NjAwMCwxMDAwLDY4MDAwLDEwMDAsNjgwMDAsMTAwMCw0MTAwMCwzODAwMCwxMTAwMCwxMDAwLDc1MDAwLDMwMDAsNTEwMDAsMTQwMDAsNzQwMDAsMjQwMDAsODQwMDAsODIwMDAsMTcwMDAsMzAwMCw1NDAwMCw3NDAwMCw0NjAwMCwxMDAwLDQwMDAwLDI5MDAwLDI5MDAwLDY3MDAwLDc4MDAwLDI1MDAwLDQ4MDAwLDIxMDAwLDIxMDAwLDE2MDAwLDc0MDAwLDY1MDAwLDIwMDAwLDQxMDAwLDM4MDAwLDI2MDAwLDM1MDAwLDMxMDAwLDkwMDAsNjcwMDAsMzMwMDAsMTAwMCw1MTAwMCwxMTYwMDAsNjkwMDAsODgwMDAsODEwMDAsNDcwMDAsNDMwMDAsMjAwMCw5ODAwMCwyNTAwMCw5MDAwLDk4MDAwLDEzMDAwLDExMDAwLDI2MDAwLDEwMDAsMjEwMDAsMTAwMDAsNTIwMDAsMjkwMDAsNDgwMDAsMTIwMDAsODMwMDAsOTcwMDAsMTAwMCw4NTAwMCwxMDUwMDAsMzgwMDAsOTAwMCwyODAwMCw1MjAwMCw5NzAwMCwxMDAwLDcyMDAwLDY0MDAwLDEwNDAwMCw3NTAwMCw5NTAwMCw2OTAwMCw4MDAwLDI4MDAwLDc4MDAwXSwib3BzIjo0NDgyMH1dLCJ1cGRhdGVkIjoiMjAyMi0wOS0yN1QwNDoxNjo1OC4yODdaIn0%3D)

image

**Comment by HeySora** [comment]
That's been done! :)  
Anything else I need to do before this can be merged?

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
:rocket:
